SIGSEGV on malformed curl

## Problem
It seems that when our dns server is running on docker, if we send it a malformed curl message (either in another docker container or locally), it crashes the dns server( I suppose this may be a fun hacky way to test fault tolerance)

## To Reproduce

run a dns server on docker 
```
make docker-build
make docker-run
```
and then from another terminal, run 
```
curl -f "google.com. IN A 10.0.0.5" -X PUT -L http://127.0.0.1:8000/add
```
(this is a small typo of a curl message. With an `-f` instead of a `-d` flag). 

This crashes the docker instance running with the following stack trace:
```
docker run --ip 172.30.100.104 \
		--publish 8000:12380 --publish 8001:53/udp \
		--expose 53/udp --expose 12380 --expose 12379 \
		dns_server:0.1 ./dns_server --id 1 --cluster http://127.0.0.1:12379 --port 12380
2020-05-14 00:24:14.321865 I | cluster ip http://127.0.0.1:12380
2020-05-14 00:24:14.322422 I | replaying WAL of member 1
2020-05-14 00:24:14.322776 I | loading WAL at term 0 and index 0
raft2020/05/14 00:24:14 INFO: 1 became follower at term 18
raft2020/05/14 00:24:14 INFO: newRaft 1 [peers: [], term: 18, commit: 25, applied: 0, lastindex: 25, lastterm: 18]
raft2020/05/14 00:24:16 INFO: 1 is starting a new election at term 18
raft2020/05/14 00:24:16 INFO: 1 became candidate at term 19
raft2020/05/14 00:24:16 INFO: 1 received MsgVoteResp from 1 at term 19
raft2020/05/14 00:24:16 INFO: 1 became leader at term 19
raft2020/05/14 00:24:16 INFO: raft.node: 1 elected leader 1 at term 19
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x18 pc=0x9edac7]

goroutine 29 [running]:
main.(*dnsStore).insertRR(0xc000064a20, 0x0, 0x0)
	/go/src/app/dns_store.go:431 +0x37
main.(*dnsStore).readCommits(0xc000064a20, 0xc000082180, 0xc0000821e0)
	/go/src/app/dns_store.go:491 +0x46a
created by main.newDNSStore
	/go/src/app/dns_store.go:65 +0x111
make: *** [docker-run] Error 2
```
I haven't touched enough of the code in dns_store.go to know if possibly this is a problem with the docker set up (the bug doesn't appear if you don't use docker for example), so any help about where to look to fix this would be appreciated.

--- a/dns/dns_store.go
+++ b/dns/dns_store.go
@@ -85,7 +85,7 @@ func stringsToRRs(sList []string) []dns.RR {
 	rrs := []dns.RR{}
 	for _, s := range sList {
 		rr, err := dns.NewRR(s)
-		if err == nil {
+		if err == nil && rr != nil {
 			rrs = append(rrs, rr)
 		}
 	}
@@ -223,7 +223,7 @@ func (s *dnsStore) getGlueRecords(nsName string) []*dns.RR {
 		if glueARecords, ok := typeMap[dns.TypeA]; ok {
 			for _, glueRecordString := range glueARecords {
 				rr, err := dns.NewRR(glueRecordString)
-				if err == nil {
+				if err == nil && rr != nil {
 					hasPreciseMatch = true
 					glueRecords = append(glueRecords, &rr)
 				}
@@ -247,7 +247,7 @@ func (s *dnsStore) getGlueRecords(nsName string) []*dns.RR {
 			wildcardGlueRecordList := wildcardTypeMap[dns.TypeA]
 			for _, wildCardGlueRecordString := range wildcardGlueRecordList {
 				rr, err := dns.NewRR(wildCardGlueRecordString)
-				if err == nil {
+				if err == nil && rr != nil {
 					hasMatch = true
 					glueRecords = append(glueRecords, &rr)
 				}
@@ -272,7 +272,7 @@ func (s *dnsStore) getCacheRecords(domainName string, qType uint16) []*dns.RR {
 				// check cache validity
 				timeDiff := uint32(time.Since(crInfo.createTime).Seconds())
 				// newTTL := validCache(&crInfo)
-				if err == nil && timeDiff < crInfo.ttl {
+				if err == nil && cr != nil && timeDiff < crInfo.ttl {
 					// restore ttl
 					cr.Header().Ttl = crInfo.ttl - timeDiff
 					cacheRecords = append(cacheRecords, &cr)
@@ -300,7 +300,7 @@ func (s *dnsStore) HandleSingleQuestion(name string, qType uint16, r *dns.Msg) b
 		if rrStringList != nil && len(rrStringList) != 0 {
 			for _, rrString := range rrStringList {
 				rr, err := dns.NewRR(rrString)
-				if err == nil {
+				if err == nil && rr != nil {
 					hasPreciseMatch = true // We have a precise match if we push entry
 					r.Answer = append(r.Answer, rr)
 				}
@@ -365,7 +365,7 @@ func (s *dnsStore) HandleSingleQuestion(name string, qType uint16, r *dns.Msg) b
 			wildcardRRList := wildcardTypeMap[qType]
 			for _, wildCardRRString := range wildcardRRList {
 				rr, err := dns.NewRR(wildCardRRString)
-				if err == nil {
+				if err == nil && rr != nil {
 					hasMatch = true
 					// Spec asks us to change the owner to be w/o star
 					rr.Header().Name = domainName
@@ -486,7 +486,7 @@ func (s *dnsStore) readCommits(commitC <-chan *string, errorC <-chan error) {
 		switch proposal.ProposalType {
 		case AddRR:
 			rr, err := dns.NewRR(proposal.RRString)
-			if err == nil {
+			if err == nil && rr != nil {
 				s.mu.Lock()
 				s.insertRR(rr)
 				s.mu.Unlock()

--- a/dns/hash_server/hash_server.go
+++ b/dns/hash_server/hash_server.go
@@ -84,7 +84,7 @@ func (c *jsonClusterInfo) intoClusterInfo() clusterInfo {
 	now := time.Now()
 	for _, m := range c.Members {
 		glueRecord, err := dns.NewRR(m.GlueRecord)
-		if err != nil {
+		if err != nil || glueRecord == nil {
 			continue
 		}
 		mi := nsInfo{
@@ -151,7 +151,7 @@ func serveHashServerHTTPAPI(store *hashServerStore, port int, done chan<- error)
 		rrString := string(body)
 
 		rr, err := dns.NewRR(rrString)
-		if err != nil {
+		if err != nil || rr == nil {
 			log.Println("Bad RR request")
 			http.Error(w, "Bad RR request", http.StatusBadRequest)
 			return

--- a/dns/httpapi.go
+++ b/dns/httpapi.go
@@ -14,8 +14,8 @@ import (
 )
 
 func checkValidRRString(s string) bool {
-	_, err := dns.NewRR(s)
-	return err == nil
+	rr, err := dns.NewRR(s)
+	return err == nil && rr != nil
 }
 
 // func (h *dnsHTTPAPI) ServeHTTP(w http.ResponseWriter, r *http.Request) {
@@ -158,7 +158,7 @@ func serveHTTPAPI(store *dnsStore, port int, confChangeC chan<- raftpb.ConfChang
 		rrString := string(body)
 
 		rr, err := dns.NewRR(rrString)
-		if err != nil {
+		if err != nil || rr == nil {
 			log.Println("Bad cache RR request")
 			http.Error(w, "Bad RR request", http.StatusBadRequest)
 			return

